File: frontend/src/entities/api/material.ts
// src/entities/api/material.ts
import { Material } from '@/entities/interfaces/material';
import { api } from './axios';

export const getMaterials = async (): Promise<Material[]> => {
  try {
    const response = await api.get('/materials');
    return response.data;
  } catch (error) {
    console.error('Failed to fetch materials:', error);
    throw error;
  }
};

export const getMaterialById = async (id: number): Promise<Material> => {
  try {
    const response = await api.get(`/materials/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Failed to fetch material with id ${id}:`, error);
    throw error;
  }
};

export const createMaterial = async (newMaterial: Material): Promise<Material> => {
  try {
    const response = await api.post('/admin/materials', newMaterial, {
      headers: {
        Authorization: localStorage.getItem('token')
      }
    });
    return response.data;
  } catch (error) {
    console.error('Failed to create material:', error);
    throw error;
  }
};

export const updateMaterial = async (id: number, updatedMaterial: Material): Promise<Material> => {
  try {
    const response = await api.put(`/admin/materials/${id}`, updatedMaterial, {
      headers: {
        Authorization: localStorage.getItem('token')
      }
    });
    return response.data;
  } catch (error) {
    console.error('Failed to update material:', error);
    throw error;
  }
};

export const deleteMaterial = async (id: number): Promise<void> => {
  try {
    await api.delete(`/admin/materials/${id}`, {
      headers: {
        Authorization: localStorage.getItem('token')
      }
    });
  } catch (error) {
    console.error('Failed to delete material:', error);
    throw error;
  }
};

File: backend/repositories/material.go
package repositories

import (
	"chukotka/core/db"
	"chukotka/models"
	"errors"
)

func GetAllMaterials() ([]models.Material, error) {
	var materials []models.Material
	result := db.DB.Find(&materials)
	return materials, result.Error
}

func GetMaterialByID(id uint) (models.Material, error) {
	var material models.Material
	result := db.DB.First(&material, id)
	if result.Error != nil {
		return material, errors.New("material not found")
	}
	return material, nil
}

func CreateMaterial(material *models.Material) error {
	return db.DB.Create(material).Error
}

func UpdateMaterial(material *models.Material) error {
	return db.DB.Save(material).Error
}

func DeleteMaterial(id uint) error {
	result := db.DB.Delete(&models.Material{}, id)
	if result.RowsAffected == 0 {
		return errors.New("material not found")
	}
	return nil
}
